### PR TITLE
Top-down view: hooked indication, hook/unhook context actions

### PR DESCRIPTION
**Pull request by dpallotti**
#### Show whether a function is hooked also in the top-down view
This is done by adding the "[✓]" string in front of the function name.
#### Add "Hook" and "Unhook" context menu actions to the top-down view
Bug: http://b/162707037

### CODE REVIEW

**Review by beckerhe — APPROVED** [review]
LGTM! Thanks!

**Comment by ronaldfw** [comment]
Two quick questions:
(1) How do users learn/know that "[✓]" means that a function is hooked? It's obvious when you use the "Hook" action in the top down view and this appears, but if a function was hooked in a different context, then it may not be obvious. The tooltip could add this information (and see (2)).
(2) Do we already use "[✓]" in other views? I think it should be consistent. If used everywhere, this also mostly removes my concerns with (1), as the "[✓]" will appear whenever a user hooks a function, no matter which UI context they are in.

**Comment by dpallotti** [comment]
> Two quick questions:
> (1) How do users learn/know that "[✓]" means that a function is hooked? It's obvious when you use the "Hook" action in the top down view and this appears, but if a function was hooked in a different context, then it may not be obvious. The tooltip could add this information (and see (2)).
> (2) Do we already use "[✓]" in other views? I think it should be consistent. If used everywhere, this also mostly removes my concerns with (1), as the "[✓]" will appear whenever a user hooks a function, no matter which UI context they are in.

Well, this is what I was mentioning yesterday. I'm not sure what the best notation is, so in the end I just went for one (the specific string can always easily be changed), but it's not that obvious at first. An alternative I tried is to use "[HOOKED]" or "[hooked]"&mdash;it doesn't look that great as it's a bit long but for sure it makes it obvious. I will at least add it to the tooltip though, I forgot about that.

Other views use a dedicated column, but I don't think we want to do the same here, as the column wouldn't be first. In those columns, the character for hooked functions is "X". We could replace it with "✓" everywhere, it might help a bit (and I guess it also makes more sense), but I don't think it really solves the problem.

**Comment by dpallotti** [comment]
> > Other views use a dedicated column, but I don't think we want to do the same here, as the column wouldn't be first. In those columns, the character for hooked functions is "X". We could replace it with "✓" everywhere, it might help a bit (and I guess it also makes more sense), but I don't think it fully solves the problem as it's still quite a generic character.
> 
> I guess we don't want to remove the dedicated column, as it still makes sense to order by hooked functions? I would at least think we should replace the "X" with "✓", as you suggest.

I'll do that, together with "[HOOKED]" in the tooltip it is quite an improvement despite my concerns in the previous message.

**Comment by dpallotti** [comment]
> I guess we don't want to remove the dedicated column, as it still makes sense to order by hooked functions? I would at least think we should replace the "X" with "✓", as you suggest.

I replaced "X" with "✓" but I also had to replace "-" with "" as the visual contrast wouldn't be enough otherwise. I think it also looks better. Either way, they are not constant and they're easy to change. If we don't want that, I could also use "[X]" in the top-down view instead.

File: OrbitGl/FunctionsDataView.cpp [hunk 1]
```diff
@@ -15,6 +15,9 @@ using orbit_client_protos::FunctionInfo;
 
 FunctionsDataView::FunctionsDataView() : DataView(DataViewType::kFunctions) {}
 
+const std::string FunctionsDataView::kSelectedFunctionString = "✓";
+const std::string FunctionsDataView::kNonSelectedFunctionString = "";
```
Comment (ronaldfw) [inline OrbitGl/FunctionsDataView.cpp:19]: Nit: Not 100% sure, but "non selected" sounds wrong to me. "Unselected", maybe?

**Review by ronaldfw — APPROVED** [review]
LGTM!

File: OrbitGl/FunctionsDataView.cpp (hunk 1 above)
Comment (dpallotti) [reply OrbitGl/FunctionsDataView.cpp]: The reasoning was that "unselected" feels like a functions was selected and then deselected, but I don't like "non selected" either, so I changed it.

File: OrbitQt/topdownwidget.cpp [hunk 2]
```diff
@@ -276,3 +303,35 @@ bool TopDownWidget::HighlightCustomFilterSortFilterProxyModel::ItemMatchesFilter
...
 }
+
+QVariant TopDownWidget::HookedIdentityProxyModel::data(const QModelIndex& index, int role) const {
+  QVariant data = QIdentityProxyModel::data(index, role);
+  if ((role != Qt::DisplayRole && role != Qt::ToolTipRole) ||
+      index.column() != TopDownViewItemModel::kThreadOrFunction) {
+    return data;
+  }
+
+  bool is_ulonglong;
```
[the remark this answers is not in the document]
Comment (dpallotti) [reply OrbitQt/topdownwidget.cpp:314]: This is passed as a pointer to and set by `toULongLong`, so I would think that conceptually it should not be initialized as it's about to be overwritten in any case. Not sure what the general practice for these cases is though.

[the remark this answers is not in the document]
Comment (ronaldfw) [reply OrbitQt/topdownwidget.cpp:314]: I would always err on the side of initializing with a safe default value. Are you okay with this being a random value? Are you okay with this being a random value if someone later changes the code below to check if this variable is ulonglong? Do you have the guarantee that toUlonglong actually writes the value, even in a potential error case?

[the remark this answers is not in the document]
Comment (dpallotti) [reply OrbitQt/topdownwidget.cpp]: Alright then, done.

**Review by florian-kuebler — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**